vyper/parser/expr.py

```diff
@@ -57,6 +61,107 @@
     "chain",
 }
 
+# Necessary to ensure we have enough precision to do the log/exp calcs
+getcontext().prec = 42
+
+
+def calculate_largest_power(a: int, num_bits: int, is_signed: bool) -> int:
+    """
+    For a given base `a`, compute the maximum power `b` that will not
+    produce an overflow in the equation `a ** b`
+
+    Arguments
+    ---------
+    a : int
+        Base value for the equation `a ** b`
+    num_bits : int
+        The maximum number of bits that the resulting value must fit in
+    is_signed : bool
+        Is the operation being performed on signed integers?
+
+    Returns
+    -------
+    int
+        Largest possible value for `b` where the result does not overflow
+        `num_bits`
+    """
+    if num_bits % 8:
+        raise CompilerPanic("Type is not a modulo of 8")
+
+    value_bits = num_bits - (1 if is_signed else 0)
+    if a >= 2 ** value_bits:
+        raise TypeCheckFailure("Value is too large and will always throw")
+    elif a < -(2 ** value_bits):
+        raise TypeCheckFailure("Value is too small and will always throw")
+
+    a = abs(a)
+    try:
+        b = int(Decimal(value_bits) / (Decimal(a).ln() / Decimal(2).ln()))
+    except (DivisionByZero, InvalidOperation):
+        return 1
+    if b <= 1:
+        return 1  # Value is assumed to be in range, therefore power of 1 is max
+
+    # Do a bit of iteration to ensure we have the exact number
+    num_iterations = 0
+    while a ** (b + 1) < 2 ** value_bits:
+        b += 1
+        num_iterations += 1
+        assert num_iterations < 10000
+    while a ** b >= 2 ** value_bits:
+        b -= 1
+        num_iterations += 1
+        assert num_iterations < 10000
+
+    return b
```

Comment:
This change will get the edge cause. Notice the new `CompilerPanic` (which you may want to handle separately by just not applying the operation and returning the un-clamped `exp` operation)
```suggestion
    a_is_negative = a < 0
    a = abs(a)  # No longer need to know if it's signed or not
    if a == 0 or a == 1:
        raise CompilerPanic("Exponential operation is useless!")
        
    # NOTE: There is an edge case if `a` were left signed where the following
    #       operation would not work (`ln(a)` is undefined if `a <= 0`)
    b = int(Decimal(value_bits) / (Decimal(a).ln() / Decimal(2).ln()))
    if b <= 1:
        return 1  # Value is assumed to be in range, therefore power of 1 is max

    # Do a bit of iteration to ensure we have the exact number
    num_iterations = 0
    while a ** (b + 1) < 2 ** value_bits:
        b += 1
        num_iterations += 1
        assert num_iterations < 10000
    while a ** b >= 2 ** value_bits:
        b -= 1
        num_iterations += 1
        assert num_iterations < 10000

    # Edge case: If a is negative and the values of a and b are such that:
    #               (a) ** (b + 1) == -(2 ** value_bits)
    #            we can actually squeak one more out of it because it's on the edge
    if a_is_negative and (-a) ** (b + 1) == -(2 ** value_bits):  # NOTE: a = abs(a)
        return b + 1
    else:
        return b  # Exact
```